#### STEP 1: Loading Dependencies

In [41]:
import pandas as pd


#### STEP 2: Load data

In [42]:
df = pd.read_csv("../data/raw/healthcare_dataset.csv")

##### STEP 3: Inspecting the dataset

In [43]:
def check_df(df):
    print("HEAD")
    display(df.head())
    
    print("\nSHAPE")
    print(df.shape)
    
    print("\nCOLUMNS")
    print(df.columns.tolist())
    
    print("\nINFO")
    df.info()
    
    print("\nMISSING VALUES")
    print(df.isnull().sum())
    
    print("\nDUPLICATES")
    print(df.duplicated().sum())

check_df(df)

HEAD


,Name,Age,Gender,Blood Type,Medical Condition,Date of Admission,Doctor,Hospital,Insurance Provider,Billing Amount,Room Number,Admission Type,Discharge Date,Medication,Test Results
0,Bobby JacksOn,30,Male,B-,Cancer,2024-01-31,Matthew Smith,Sons and Miller,Blue Cross,18856.281306,328,Urgent,2024-02-02,Paracetamol,Normal
1,LesLie TErRy,62,Male,A+,Obesity,2019-08-20,Samantha Davies,Kim Inc,Medicare,33643.327287,265,Emergency,2019-08-26,Ibuprofen,Inconclusive
2,DaNnY sMitH,76,Female,A-,Obesity,2022-09-22,Tiffany Mitchell,Cook PLC,Aetna,27955.096079,205,Emergency,2022-10-07,Aspirin,Normal
3,andrEw waTtS,28,Female,O+,Diabetes,2020-11-18,Kevin Wells,"Hernandez Rogers and Vang,",Medicare,37909.782410,450,Elective,2020-12-18,Ibuprofen,Abnormal
4,adrIENNE bEll,43,Female,AB+,Cancer,2022-09-19,Kathleen Hanna,White-White,Aetna,14238.317814,458,Urgent,2022-10-09,Penicillin,Abnormal



SHAPE
(55500, 15)

COLUMNS
['Name', 'Age', 'Gender', 'Blood Type', 'Medical Condition', 'Date of Admission', 'Doctor', 'Hospital', 'Insurance Provider', 'Billing Amount', 'Room Number', 'Admission Type', 'Discharge Date', 'Medication', 'Test Results']

INFO
<class 'pandas.DataFrame'>
RangeIndex: 55500 entries, 0 to 55499
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Name                55500 non-null  str    
 1   Age                 55500 non-null  int64  
 2   Gender              55500 non-null  str    
 3   Blood Type          55500 non-null  str    
 4   Medical Condition   55500 non-null  str    
 5   Date of Admission   55500 non-null  str    
 6   Doctor              55500 non-null  str    
 7   Hospital            55500 non-null  str    
 8   Insurance Provider  55500 non-null  str    
 9   Billing Amount      55500 non-null  float64
 10  Room Number         55500 non-null  int64  
 11  A

In [44]:
# Investigating the target variable distribution

df["Test Results"].value_counts()

Test Results
Abnormal        18627
Normal          18517
Inconclusive    18356
Name: count, dtype: int64

In [45]:
df["Test Results"].value_counts(normalize=True)

Test Results
Abnormal        0.335622
Normal          0.333640
Inconclusive    0.330739
Name: proportion, dtype: float64

##### Target Distribution

The target variable `Test Results` contains three classes:

- Abnormal: 18,627
- Normal: 18,517
- Inconclusive: 18,356

The classes are approximately evenly distributed, indicating that the target is balanced. 

In [46]:
# Inspecting Unique values in categorical columns
categorical_columns = df.select_dtypes(include="str").columns

for column in categorical_columns:
    print(f"\n{column}")
    print(df[column].nunique())
    print(df[column].unique()[:20])


Name
49992
<StringArray>
[      'Bobby JacksOn',        'LesLie TErRy',         'DaNnY sMitH',
        'andrEw waTtS',       'adrIENNE bEll',       'EMILY JOHNSOn',
      'edwArD EDWaRDs',  'CHrisTInA MARtinez',     'JASmINe aGuIlaR',
    'ChRISTopher BerG',    'mIchElLe daniELs',      'aaRon MARtiNeZ',
       'connOR HANsEn',        'rObeRt bAuer',        'bROOkE brady',
  'MS. nAtalIE gAMble',       'haley perkins', 'mRS. jamiE cAMPBELl',
        'LuKE BuRgEss',      'dANIEL schmIdt']
Length: 20, dtype: str

Gender
2
<StringArray>
['Male', 'Female']
Length: 2, dtype: str

Blood Type
8
<StringArray>
['B-', 'A+', 'A-', 'O+', 'AB+', 'AB-', 'B+', 'O-']
Length: 8, dtype: str

Medical Condition
6
<StringArray>
['Cancer', 'Obesity', 'Diabetes', 'Asthma', 'Hypertension', 'Arthritis']
Length: 6, dtype: str

Date of Admission
1827
<StringArray>
['2024-01-31', '2019-08-20', '2022-09-22', '2020-11-18', '2022-09-19',
 '2023-12-20', '2020-11-03', '2021-12-28', '2020-07-01', '2021-05-23',
 '2020-0

In [ ]:
# Inspecting numerical columns
numerical_columns = df.select_dtypes(include=["int64", "float64"]).columns

numerical_columns


Index(['Age', 'Billing Amount', 'Room Number'], dtype='str')

In [48]:
df[["Age", "Billing Amount", "Room Number"]].describe()

,Age,Billing Amount,Room Number
count,55500.000000,55500.000000,55500.000000
mean,51.539459,25539.316097,301.134829
std,19.602454,14211.454431,115.243069
min,13.000000,-2008.492140,101.000000
25%,35.000000,13241.224652,202.000000
50%,52.000000,25538.069376,302.000000
75%,68.000000,37820.508436,401.000000
max,89.000000,52764.276736,500.000000


- Negative minimum billing amount(investigate)

In [49]:
df[df["Billing Amount"] < 0][["Billing Amount", "Test Results"]]

,Billing Amount,Test Results
132,-502.507813,Normal
799,-1018.245371,Inconclusive
1018,-306.364925,Normal
1421,-109.097122,Abnormal
2103,-576.727907,Abnormal
...,...,...
52894,-353.865186,Inconclusive
53204,-306.364925,Normal
53232,-591.917419,Abnormal
54136,-199.663795,Abnormal


In [62]:
# Inspecting the billing amount column
print("Minimum:", df["Billing Amount"].min())
print("Maximum:", df["Billing Amount"].max())

print("\nZero billing amounts:")
print((df["Billing Amount"] == 0).sum())

print("\nNegative billing amounts:")
print((df["Billing Amount"] < 0).sum())

print("\nPositive billing amounts:")
print((df["Billing Amount"] > 0).sum())

Minimum: -2008.4921398591305
Maximum: 52764.276736469175

Zero billing amounts:
0

Negative billing amounts:
108

Positive billing amounts:
55392


In [63]:
negative_billing = df[df["Billing Amount"] < 0]

negative_billing[
    [
        "Name",
        "Age",
        "Gender",
        "Medical Condition",
        "Date of Admission",
        "Admission Type",
        "Billing Amount",
        "Test Results"
    ]
].head(10)

,Name,Age,Gender,Medical Condition,Date of Admission,Admission Type,Billing Amount,Test Results
132,ashLEy ERIcKSoN,32,Female,Cancer,2019-11-05,Urgent,-502.507813,Normal
799,CHRisTOPHer wEiss,49,Female,Asthma,2023-02-16,Elective,-1018.245371,Inconclusive
1018,AsHley WaRnER,60,Male,Hypertension,2021-12-21,Elective,-306.364925,Normal
1421,JAY galloWaY,74,Female,Asthma,2021-01-20,Emergency,-109.097122,Abnormal
2103,josHUa wilLIamSon,72,Female,Diabetes,2021-03-21,Urgent,-576.727907,Abnormal
2696,Scott VaZqUEz,74,Male,Diabetes,2023-04-12,Elective,-135.986000,Abnormal
2855,CaROl aNDErSoN,39,Female,Hypertension,2020-04-03,Elective,-370.983674,Abnormal
3772,mr. ChRIStOPhER aLvARaDO,77,Male,Obesity,2022-06-03,Elective,-1310.272895,Inconclusive
5445,ALExAndrA KHaN,32,Male,Arthritis,2022-07-14,Elective,-692.408820,Abnormal
5708,JosEPh cOx,23,Male,Diabetes,2019-10-13,Elective,-353.865186,Inconclusive


In [65]:
negative_billing["Billing Amount"].value_counts().head(10)

Billing Amount
-306.364925     2
-1310.272895    2
-353.865186     2
-26.112523      2
-308.584269     2
-676.852502     2
-68.319621      2
-887.024219     2
-1316.618581    2
-591.917419     2
Name: count, dtype: int64

In [55]:
negative_billing.shape

(108, 15)

In [56]:
negative_billing["Billing Amount"].describe()

count     108.000000
mean     -499.301242
std       426.463202
min     -2008.492140
25%      -792.084802
50%      -369.093816
75%      -153.079565
max       -23.866729
Name: Billing Amount, dtype: float64

In [57]:
df.duplicated().sum()

np.int64(534)

In [58]:
duplicates = df[df.duplicated(keep=False)].sort_values(
    by=list(df.columns)
)

duplicates.head(10)

,Name,Age,Gender,Blood Type,Medical Condition,Date of Admission,Doctor,Hospital,Insurance Provider,Billing Amount,Room Number,Admission Type,Discharge Date,Medication,Test Results
42407,ABIgaIL YOung,41,Female,O+,Hypertension,2022-12-15,Edward Kramer,Moore-Mcdaniel,UnitedHealthcare,1983.568297,192,Elective,2023-01-13,Ibuprofen,Normal
54285,ABIgaIL YOung,41,Female,O+,Hypertension,2022-12-15,Edward Kramer,Moore-Mcdaniel,UnitedHealthcare,1983.568297,192,Elective,2023-01-13,Ibuprofen,Normal
26025,ALIcia taYLoR,78,Male,O+,Asthma,2022-09-18,Dawn Burton,Wright LLC,Aetna,31465.274979,149,Elective,2022-10-15,Aspirin,Inconclusive
53104,ALIcia taYLoR,78,Male,O+,Asthma,2022-09-18,Dawn Burton,Wright LLC,Aetna,31465.274979,149,Elective,2022-10-15,Aspirin,Inconclusive
42323,AMy GREEN,79,Female,B+,Obesity,2021-03-30,Brett Johnson,Taylor-Williamson,UnitedHealthcare,23402.358491,249,Elective,2021-04-27,Penicillin,Abnormal
50151,AMy GREEN,79,Female,B+,Obesity,2021-03-30,Brett Johnson,Taylor-Williamson,UnitedHealthcare,23402.358491,249,Elective,2021-04-27,Penicillin,Abnormal
21675,ANDREA HansEN,61,Male,O+,Cancer,2021-07-02,Alisha Flores,LLC Clark,Cigna,40026.763948,254,Elective,2021-07-22,Paracetamol,Normal
51695,ANDREA HansEN,61,Male,O+,Cancer,2021-07-02,Alisha Flores,LLC Clark,Cigna,40026.763948,254,Elective,2021-07-22,Paracetamol,Normal
36207,ANDrEA fREnCH,73,Male,A-,Arthritis,2021-02-08,Danny Oconnor,"Taylor Small, Lin and",UnitedHealthcare,9981.590235,323,Elective,2021-02-24,Ibuprofen,Inconclusive
51916,ANDrEA fREnCH,73,Male,A-,Arthritis,2021-02-08,Danny Oconnor,"Taylor Small, Lin and",UnitedHealthcare,9981.590235,323,Elective,2021-02-24,Ibuprofen,Inconclusive


In [59]:
duplicates.shape

(1068, 15)

In [60]:
duplicate_counts = df.value_counts()

duplicate_counts.head(10)

Name             Age  Gender  Blood Type  Medical Condition  Date of Admission  Doctor             Hospital                     Insurance Provider  Billing Amount  Room Number  Admission Type  Discharge Date  Medication   Test Results
Nancy glOVeR     58   Male    A-          Hypertension       2020-05-08         Jennifer Larson    Khan, and Rodriguez Fischer  Medicare            19183.168885    378          Emergency       2020-06-01      Aspirin      Abnormal        2
DAVid higgInS    49   Female  B-          Arthritis          2021-03-05         Erin Henderson MD  Evans and Hall Schneider,    Medicare            24948.477824    361          Emergency       2021-03-20      Penicillin   Abnormal        2
RoBErt hIGGInS   42   Male    AB-         Asthma             2021-05-06         Scott Davis        and Ford Lee, Rodriguez      Medicare            13355.782085    451          Elective        2021-05-29      Ibuprofen    Inconclusive    2
kevIn HiCKs      66   Male    AB+         Art

In [61]:
duplicate_counts.value_counts().sort_index()

count
1    54432
2      534
Name: count, dtype: int64

In [66]:
# Investigating time-related columns
admission_dates = pd.to_datetime(df["Date of Admission"], errors="coerce")
discharge_dates = pd.to_datetime(df["Discharge Date"], errors="coerce")

In [ ]:
# Check for empty dates
print("Invalid admission dates:", admission_dates.isna().sum())
print("Invalid discharge dates:", discharge_dates.isna().sum())

Invalid admission dates: 0
Invalid discharge dates: 0


In [68]:
length_of_stay = discharge_dates - admission_dates

length_of_stay.describe()

count                      55500
mean     15 days 12:12:58.378378
std       8 days 15:49:49.456115
min              1 days 00:00:00
25%              8 days 00:00:00
50%             15 days 00:00:00
75%             23 days 00:00:00
max             30 days 00:00:00
dtype: object

In [69]:
print("Negative stays:", (length_of_stay < pd.Timedelta(0)).sum())
print("Same-day stays:", (length_of_stay == pd.Timedelta(0)).sum())

Negative stays: 0
Same-day stays: 0


#### Findings

- 55,500 rows
- 15 columns
- No missing values
- 534 exact duplicate rows ; remove one copy from each pair
- 108 negative billing amounts
- No zero billing amounts
- Dates are valid and admission/discharge ordering is consistent throughout
- Target Test Results has 3 reasonably balanced classes
- Name column has to be cleaned